examples/pennylane_run_qaoa_maxcut.py

```diff
@@ -0,0 +1,287 @@
+# a measurement of multiple qubits in the computational basis, so we define
+# a Hermitian operator to do this. The eigenvalues of the operator are
+# the joint qubit measurement values in integer form.
+
+
+def comp_basis_measurement(wires):
+    n_wires = len(wires)
+    return qml.Hermitian(np.diag(range(2 ** n_wires)), wires=wires)
+
+
+##############################################################################
+# Circuit
+# ~~~~~~~
+# Next, we create a quantum device with 4 qubits.
+
+dev1 = qml.device("default.qubit", wires=n_wires, analytic=True, shots=1)
+
+##############################################################################
+# We also require a quantum node which will apply the operators according to the
+# angle parameters, and return the expectation value of the observable
+# :math:`\sigma_z^{j}\sigma_z^{k}` to be used in each term of the objective function later on. The
+# argument ``edge`` specifies the chosen edge term in the objective function, :math:`(j,k)`.
+# Once optimized, the same quantum node can be used for sampling an approximately optimal bitstring
+# if executed with the ``edge`` keyword set to ``None``. Additionally, we specify the number of layers
+# (repeated applications of :math:`U_BU_C`) using the keyword ``n_layers``.
+
+pauli_z = [[1, 0], [0, -1]]
+pauli_z_2 = np.kron(pauli_z, pauli_z)
+
+
+@qml.qnode(dev1)
+def circuit(params, edge=None, n_layers=1):
+    # apply hadamards to get n qubit |+> state
+    for wire in range(n_wires):
+        qml.Hadamard(wires=wire)
+    # p instances of unitary operators
+    for i in range(n_layers):
+        U_C(params[0][i]) # params[0] is gamma
+        U_B(params[1][i]) # params[1] is beta
+    if edge is None:
+        # measurement phase
+        return qml.sample(comp_basis_measurement(range(n_wires)))
+    # during the optimization phase we are evaluating a term
+    # in the objective using exp val
+    return qml.expval(qml.Hermitian(pauli_z_2, wires=edge))
+
+
+##############################################################################
+# Optimization
+# ~~~~~~~~~~~~
+# Finally, we optimize the objective over the
+# angle parameters :math:`\boldsymbol{\gamma}` (``params[0]``) and :math:`\boldsymbol{\beta}`
+# (``params[1]``).
+# and then sample the optimized
+# circuit multiple times to yield a distribution of bitstrings. One of the optimal partitions
+# (:math:`z=0101` or :math:`z=1010`) should be the most frequently sampled bitstring.
+
+
+def qaoa_maxcut(n_layers=1):
+    print("\np={:d}".format(n_layers))
+
+    # initialize the parameters near zero
+    init_params = 0.01 * np.random.rand(2, n_layers)
+
+    # minimize negative the objective
+    def objective(params):
+        neg_obj = 0
+        for edge in graph:
+            # objective for the maxcut problem
+            neg_obj -= 0.5 * (1 - circuit(params, edge=edge, n_layers=n_layers))
+        return neg_obj
+
+    # initialize optimizer
```

Comment:
Why this optimizer? It's nice to explain choices in tutorials

Reply:
Well, any gradient-based optimizer would work also - should I still include an explanation for choosing gradient descent for example?

Reply:
I've actually left the optimizer as is and will include some justification, but in reality I just played around and found Adagrad worked better

Reply:
please add one-line justification (can just be by observation that optimizer seems to work well)